src/rust/runtime/memory/demibuffer.rs

```diff
@@ -428,52 +428,106 @@ impl DemiBuffer {
+    /// a [Fail] structure encoding the failure condition is returned instead.
+    ///
+    /// **Notes**
+    ///
+    /// - The target [DemiBuffer] must be a single buffer segment (not a chain).
+    /// - The target [DemiBuffer] should be large enough to hold `offset`.
+    ///
+    pub fn split_back(&mut self, offset: usize) -> Result<Self, Fail> {
+        // Check if this is a multi-segment buffer.
+        if self.is_multi_segment() {
+            let cause: String = format!("cannot split a multi-segment buffer");
+            error!("split_back(): {}", &cause);
+            return Err(Fail::new(libc::EINVAL, &cause));
+        }
+
+        // Check if split offset is valid.
+        if self.len() < offset {
+            let cause: String = format!("cannot split buffer at given the offset (offset={:?})", offset);
+            error!("split_back(): {}", &cause);
+            return Err(Fail::new(libc::EINVAL, &cause));
         }
 
-        // Clone ourselves.
+        // Clone the target buffer before any changes are applied.
         let mut back_half: DemiBuffer = self.clone();
 
-        // Remove data starting at `offset` from the front (original) DemiBuffer as those bytes now belong to the back.
-        let trim: usize = self.len() - offset;
+        // Remove data starting at `offset` from the front half buffer (original buffer).
+        // Those bytes now belong to the back half buffer.
         // This unwrap won't panic as we already performed its error checking above.
-        self.trim(trim).unwrap();
+        self.trim(self.len() - offset).unwrap();
 
-        // Remove `offset` bytes from the beginning of the back (clone) DemiBuffer as they now belong to the front.
+        // Remove `offset` bytes from the beginning of the back half buffer (cloned buffer).
+        // Those bytes now belong to the front back buffer.
         // This unwrap won't panic as we already performed its error checking above.
         back_half.adjust(offset).unwrap();
 
-        // Return the back DemiBuffer.
+        // Return the back half buffer.
         Ok(back_half)
     }
 
+    ///
+    /// **Description**
+    ///
+    /// Splits the target [DemiBuffer] at the given `offset` and returns a new [DemiBuffer] containing the data before
+    /// the split point (front half).
+    ///
+    /// The data contained in the new [DemiBuffer] is removed from the original [DemiBuffer] (back half).
+    ///
+    /// **Return Value**
+    ///
+    /// On successful completion, a new [DemiBuffer] containing the data before the split point is returned. On failure,
+    /// a [Fail] structure encoding the failure condition is returned instead.
+    ///
+    /// **Notes**
+    ///
+    /// - The target [DemiBuffer] must be a single buffer segment (not a chain).
+    /// - The target [DemiBuffer] should be large enough to hold `offset`.
+    ///
+    pub fn split_front(&mut self, offset: usize) -> Result<Self, Fail> {
```

Comment:
To eliminate duplicate code, we should probably have a shared function here that gives both halves back and then split_front and split_back just return the front_half or back_half.

Reply:
Sure... I merged the two functions in one, called `split()`